### ISSUE TITLE
EE3 Compatibility

### ISSUE THREAD

**Issue by westhouseit** [westhouseit/Postmaster#1]
Looking for feedback from those interested in EE3 compatibility. Perhaps this is something we can crowdfund.

**Comment by kemmerly** [comment]
I would be interested in this. Have a site that needs to be upgraded and this is a key integration.

**Comment by ghost** [comment]
I'd be interested as well.

**Comment by westhouseit** [comment]
Are either of you developers or have access to a developer who can contribute some time towards the project?

**Comment by dankingsley** [comment]
I'd also be really interested, but unfortunately am not a developer! Potentially up for partially helping to fund it though if needs be. Have a couple of fairly key sites using it currently.

**Comment by thkro** [comment]
I am also very interested, happy to fund this too

**Comment by 620studio** [comment]
I'm getting estimates from developers to convert PM to EE3. Because of legality purposes (https://expressionengine.com/forums/topic/249380/ee2-reprieve-ee3-third-party-plugin-support) the developer wouldn't be able to sell the new version as they don't own the original code. I've tried on multiple occasions to contact ObjectiveHTML about completely open-sourcing the project or at least allowing someone to acquire the rights but I get no response. Maybe the new developer would have better luck.

The only way I could see us getting a distributed EE3 version of PostMaster would be to pay for the conversion and make it free to download. Unfortunately, the continued support would be an issue as I can't imagine any developer would be willing to a) support a plugin they didn't originally create and b) maintain for free.

I'll let you know what kind of numbers I get back. This is a key component to my client's site as well and until PostMaster gets the EE3 treatment, we'll have to stay on EE2.

**Comment by kemmerly** [comment]
I would be willing to contribute to this - hopefully others will too. 

I have a large site that needs to be upgrade to EE3 and Postmaster is the biggest roadblock. Thanks.

**Comment by ghost** [comment]
Thanks for the update. In the same boat here. Would love to see it crowd funded into an open source project.

**Comment by westhouseit** [comment]
@620studio you're right about the developer not having full ownership of the addon, though they could have ownership of their upgrades/changes. I still think it would be cheaper for those of us/clients with Postmaster licenses to pay for the upgrade development costs and future bugfixes rather than develop a completely new addon.

@mjatharvest unfortunately to crowdfund it to open-source we'd need to either have the original developer properly open-source his code or have a new addon built from scratch. With a license for the current Postmaster I can't see any legal reason preventing code modification that makes it EE3 compatible, though each business would need to make the decision for themselves.

**Comment by dankingsley** [comment]
Also still very interested in Postmaster being upgraded to EE3, so depending on where things get to, I'd be up for helping to fund if required.

**Comment by bluebloc** [comment]
Just wondering if there have been any updates on this?

**Comment by ghost** [comment]
Looks like no reply from the developer. I forked the repo and pushed up some changes to get basic functionality running on EE3. It's probably not mature enough to make a Pull Request.

**Comment by westhouseit** [comment]
@620studio, any word from the developers you got in touch with for pricing?

@mjatharvest, great news that you've made some progress. Since I'd consider the EE3 branch to be in alpha state I'd be happy to merge in what you've got so far so we can build on it rather than redo work. How does sound putting together a list where we can track what functionality works or doesn't?

**Comment by ghost** [comment]
@westhouseit that sounds good. I've been working on this out of necessity. I can't commit to fixing the whole add-on, but will be happy to share what I've done.
[https://github.com/mjatharvest/Postmaster/commit/4b5b7150ea9f9e11d1189219d61ab352a13ffa2c](url)

**Comment by bluebloc** [comment]
I'd be happy to share the workload also.

**Comment by ghost** [comment]
The major edits made to the code base so far are:

1. use ee() instead of $this->EE
2. use a workaround to overwrite ee()->TMPL https://expressionengine.com/forums/topic/248431/override-loader-objects
3. fix the cp_url function to access settings in the back end
4. change the main extension hooks to before_channel_entry_save and after_channel_entry_save (the previous hook was deprecated)
![screen shot 2017-11-20 at 8 26 04 am](https://user-images.githubusercontent.com/285877/33023087-b5ce1d80-cdcc-11e7-9c66-ac3caa184fd2.png)

**Comment by westhouseit** [comment]
@mjatharvest could you create a pull request that I can pull into the EE3 alpha branch?

@bluebloc and others, I've created a project area and as we create issues they'll be added to the To Do column. Anyone, even those not doing development can add an Issue for a piece of work that needs doing or something that needs fixing. If you're fixing something yourself then please make a note in the issue so we don't duplicate work. Then submit a pull request when you're done.

Thanks!

**Comment by ghost** [comment]
Pull request created - I'm a bit new at this so let me know if I did anything incorrectly :)

**Comment by westhouseit** [comment]
Not a problem, pull requests is new for me too. It looks like there some merge issues. Could you please recreate the pull request with your repo having the latest changes. The **Comparing Changes** screen when creating a new pull request should say _Able to merge_.

**Comment by ghost** [comment]
Let's try the again. New PR created.

**Comment by westhouseit** [comment]
:thumbsup:

**Comment by joshconner** [comment]
Is this add-on in a working state to install on an EE3 site? The site I have has very basic needs from this add-on but I need it updated before I upgrade to EE3.

**Comment by ghost** [comment]
@joshconner I got Postmaster Parcels working out of necessity on EE3, but the upgrade process wasn't super smooth - had to change the extension hook names manually in the database. I'm happy to share with you what I have, with the understanding that it is "as is" software

**Comment by joshconner** [comment]
@mjatharvest What does "wasn't super smooth" exactly mean? If it's just manually updating a few things in the db that isn't too bad.

**Comment by ghost** [comment]
@joshconner Yes, the purpose of the comment is to set the expectation that you can't just click "install" or "update" from with the CP.

**Comment by joshconner** [comment]
@mjatharvest Awesome, I'd love to see what you have!

**Comment by ghost** [comment]
@joshconner sure thing, you can download this fork and give it a go https://github.com/mjatharvest/Postmaster or you can contact me via our site harvestmedia.com

**Comment by 620studio** [comment]
@mjatharvest Thanks for the work you've put into this so far! 

Sorry for the lack of response, I wasn't getting notified that comments were being added to this thread. The original estimate I received for doing a EE3 conversion for Postmaster was around the $6K mark. My client uses Postmaster for an annual fundraising which was around the time I was getting estimates. I've already given them advanced warning that we'll need to upgrade from EE2 over the next year which means we'll need to do something about Postmaster.

I'm glad to hear Parcels are working for EE3. We use Parcels, Hooks, and Notifications for the system so it gets us part of the way there. I'll be looking at upgrading the client starting in the beginning of hte year so I'll try getting some additional estimates for the conversion. I'll post back as soon as I hear anything.

**Comment by ghost** [comment]
Thanks @620studio. Certainly keep us updated.

**Comment by dankingsley** [comment]
Hi, I assume there are no more updates relating to this at all? Long shot I know...

**Comment by ghost** [comment]
@dankingsley sadly no

**Comment by 620studio** [comment]
As crazy as it sounds, I may have some information soon. I've been in steady communication with several dev houses who are aware of the situation. I don't have anything concrete just yet, but hope to hear something soon. Definitely trying not to give false hope, but I know there are still those of us out there desperately needing PM to be EE3+. I'll keep you posted.

**Comment by ghost** [comment]
@620studio sounds great, please do!
